File: KoohiiAite/src/main/java/com/diarmaidlindsay/koohii/model/HeisigKanji.java
package com.diarmaidlindsay.koohii.model;

import java.util.ArrayList;
import java.util.List;

/**
 * Represent an entry in the heisig_kanji table
 */
public class HeisigKanji {
    private int id;
    private String kanji;
    private boolean joyo;

    public HeisigKanji(int id, String kanji, boolean joyo) {
        this.id = id;
        this.kanji = kanji;
        this.joyo = joyo;
    }

    public HeisigKanji(int id, String kanji, int joyo) {
        this.id = id;
        this.kanji = kanji;
        this.joyo = joyo!=0;
    }

    public int getId() {
        return id;
    }

    public String getKanji() {
        return kanji;
    }

    public int getJoyo()
    {
        return joyo ? 1 : 0;
    }

    public boolean isJoyo() {
        return joyo;
    }

    public static String[] getIds(List<HeisigKanji> list)
    {
        List<String> ids = new ArrayList<>();

        for(HeisigKanji hk : list)
        {
            ids.add(String.valueOf(hk.getId()));
        }

        return ids.toArray(new String[ids.size()]);
    }

    public static List<HeisigKanji> getObjects(List<Integer> ids, List<HeisigKanji> masterList)
    {
        List<HeisigKanji> filteredList = new ArrayList<>();

        for(Integer id : ids)
        {
            filteredList.add(masterList.get(id));
        }

        return filteredList;
    }
}

File: app/src/main/java/tech/diarmaid/koohiiaite/database/entity/HeisigKanji.kt
package tech.diarmaid.koohiiaite.database.entity

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import java.util.*

/**
 * Represent an entry in the heisig_kanji table
 */
@Entity(tableName = "heisig_kanji")
data class HeisigKanji(
        @PrimaryKey @ColumnInfo(name = "id") var heisigId: Int = 0,
        var kanji: String = "",
        var joyo: Boolean = false) {

    companion object {

        /**
         * They are stored 1 indexed in the database.
         * This should be used for display purposes.
         *
         * If they are to be used for indexing other tables should
         * get the 0 Indexed list of ids instead.
         */
        fun getIds1Indexed(list: List<HeisigKanji>): List<String> {
            val ids = ArrayList<String>()

            for (hk in list) {
                ids.add(hk.heisigId.toString())
            }

            return ids
        }

        fun getHeisigKanjiMatchingIds(ids: MutableList<Int>, masterList: List<HeisigKanji>): List<HeisigKanji> {
            val filteredList = ArrayList<HeisigKanji>()
            ids.sort()

            for (id in ids) {
                //convert to 0 notation when referencing java array
                filteredList.add(masterList[id - 1])
            }

            return filteredList
        }

        /**
         * Return 4 digit Heisig Frame number for display
         */
        fun getHeisigIdAsString(heisigId: Int): String {
            var prefixZeros = ""

            if (heisigId < 1000) {
                prefixZeros += "0"
                if (heisigId < 100) {
                    prefixZeros += "0"
                    if (heisigId < 10) {
                        prefixZeros += "0"
                    }
                }
            }

            return prefixZeros + heisigId

            // 1 - 0001
            // 10 - 0010
            // 100 - 0100
            // 1000 - 1000
        }
    }
}
